File: ios/LastMinuteLive/LastMinuteLive/Core/TicketsCache.swift
import Foundation

struct CachedTicket: Codable, Equatable, Identifiable {
  var id: String { jti }
  let jti: String
  let token: String
  let orderId: String
  let performanceId: String
  let seatId: String
  let tenantId: String
  let issuedAt: Int
  let expiresAt: Int?
}

final class TicketsCacheService {
  private let keyPrefix = "tickets_cache_"
  private func key(orgId: String) -> String { keyPrefix + orgId }

  func list(orgId: String) -> [CachedTicket] {
    guard let data = KeychainHelper.load(key: key(orgId: orgId)) else { return [] }
    return (try? JSONDecoder().decode([CachedTicket].self, from: data)) ?? []
  }

  func save(orgId: String, tickets: [CachedTicket]) { if let data = try? JSONEncoder().encode(tickets) { _ = KeychainHelper.save(key: key(orgId: orgId), data: data) } }

  func upsert(orgId: String, ticket: CachedTicket) {
    var arr = list(orgId: orgId)
    if let idx = arr.firstIndex(where: { $0.jti == ticket.jti }) { arr[idx] = ticket } else { arr.append(ticket) }
    save(orgId: orgId, tickets: arr)
  }

  func remove(orgId: String, jti: String) {
    var arr = list(orgId: orgId)
    arr.removeAll { $0.jti == jti }
    save(orgId: orgId, tickets: arr)
  }
}

File: android/app/src/main/java/com/thankful/app/data/TicketsCache.kt
package com.thankful.app.data

import android.content.Context
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKeys
import org.json.JSONArray
import org.json.JSONObject

data class CachedTicket(
    val jti: String,
    val token: String,
    val orderId: String,
    val performanceId: String,
    val seatId: String,
    val tenantId: String,
    val issuedAt: Long,
    val expiresAt: Long?
)

class TicketsCacheService(ctx: Context) {
    private val masterKey = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC)
    private val prefs = EncryptedSharedPreferences.create(
        "tickets_cache",
        masterKey,
        ctx,
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
    )

    private fun key(orgId: String) = "tickets_" + orgId

    fun list(orgId: String): List<CachedTicket> {
        val json = prefs.getString(key(orgId), "[]") ?: "[]"
        val arr = JSONArray(json)
        val out = mutableListOf<CachedTicket>()
        for (i in 0 until arr.length()) {
            val o = arr.getJSONObject(i)
            out.add(
                CachedTicket(
                    jti = o.getString("jti"),
                    token = o.getString("token"),
                    orderId = o.getString("orderId"),
                    performanceId = o.getString("performanceId"),
                    seatId = o.getString("seatId"),
                    tenantId = o.getString("tenantId"),
                    issuedAt = o.getLong("issuedAt"),
                    expiresAt = if (o.has("expiresAt")) o.getLong("expiresAt") else null
                )
            )
        }
        return out
    }

    fun save(orgId: String, tickets: List<CachedTicket>) {
        val arr = JSONArray()
        tickets.forEach { t ->
            val o = JSONObject()
                .put("jti", t.jti)
                .put("token", t.token)
                .put("orderId", t.orderId)
                .put("performanceId", t.performanceId)
                .put("seatId", t.seatId)
                .put("tenantId", t.tenantId)
                .put("issuedAt", t.issuedAt)
            if (t.expiresAt != null) o.put("expiresAt", t.expiresAt)
            arr.put(o)
        }
        prefs.edit().putString(key(orgId), arr.toString()).apply()
    }

    fun upsert(orgId: String, ticket: CachedTicket) {
        val list = list(orgId).toMutableList()
        val idx = list.indexOfFirst { it.jti == ticket.jti }
        if (idx >= 0) list[idx] = ticket else list.add(ticket)
        save(orgId, list)
    }

    fun remove(orgId: String, jti: String) {
        val list = list(orgId).filterNot { it.jti == jti }
        save(orgId, list)
    }
}


